### ISSUE TITLE
Debug applications on Kube from desktop IDE

### ISSUE THREAD

**Issue by rtaniwa** [eclipse/codewind#1990]
**Codewind version:**
**OS:**

**Che version:**
**IDE extension version:**
**IDE version:**
**Kubernetes cluster:**

**Description of the enhancement:**

**Proposed solution:**

**Comment by rtaniwa** [comment]
/kind/epic

**Comment by jopit** [comment]
If we're going to require the user have kubectl on their machine, we're going to have to document that requirement.  Adding the docs label.

**Comment by jcockbain** [comment]
I've picked up the Portal side of this work.

**Comment by jcockbain** [comment]
I've picked up the Portal side of this work. 

As part of it, I'll wrap the `/project/restart` API in cwctl, as the IDEs are still calling the PFE API directly. I believe that'll be needed to make that call on a remote PFE.

In terms of running debuggers on k8s apps, there's plenty of articles around (like the ones above) that do this. https://www.ibm.com/cloud/blog/debugging-kubernetes-nodejs-app-on-ibm-cloud is one for a node app and the vscode debugger. 

So, when a Liberty / Spring / Node k8s project is restarted in debug mode (i.e. a `cwctl project restart --id <projectID>` we should be able to restart the application with a debug port (much the same as local). Then, internally call `kubectl port forward <podname> remote-port:local-port` to expose an available port locally and pass that port back as the return value.

**Comment by jcockbain** [comment]
It's relatively simple to debug a Codewind style project using port-forwarding. 

## For a Codewind style Node template

Modify dockerfile to use npm run debug.

i.e.

instead of `use ["npm", "run", "start"]`, `use ["npm", "run", "debug"]`

Project rebuilds, exposes debug port (9229) which is set in its debug script in its package.json.

Run the following command in a terminal, which will open up a port-forward session.

```shell
kubectl port-forward <pod-name> <port>
```

```javascript
{
    "type": "node",
		"name": "Debug james",
		"request": "attach",
		"port": 9229,
		"localRoot": "/Users/james.cockbain/Documents/workspace/james",
		"remoteRoot": "/app",
		"restart": true
}
```

Run debugger, can set breakpoints etc.. and hit those.

## Remaining work to automate this

- Determine way of restarting Liberty and Spring in debug.

- Find way of restarting Node, Liberty, Spring apps in "debug" mode in Kubernetes, through the PFE restart API. This currently returns a "cannot restart in Kubernetes" error, which comes from Turbine here https://github.com/eclipse/codewind/blob/master/src/pfe/file-watcher/server/src/projects/actions.ts#L296.

- Setup kubectl port-forward through cwctl, to run in the background if possible.

**Comment by jcockbain** [comment]
Wrapping restart API in cwctl on this branch -> https://github.com/jcockbain/codewind-installer/tree/add-project-restart-command

**Comment by jcockbain** [comment]
I'm working on the cwctl command that'll port-forward from a remote project pod to local. I have the basic functionality working, i.e it  will forward the debug port from a cw project running in debug mode and point the vscode debugger at it. 

This is the work still to do before I can PR and let the IDE team work against it: 

### Work to do 

- Go from projectID to k8s pod info. At the moment I'm working with hardcoded pod information, need use the projectID to look up the project pod. Either passed back from Turbine after the restart or using the go k8s client (latter should be simple to implement as we add the projectID as a label in the deployment).

- Write logic to re-portforward if connection to pods drops (this can happen quite often for some pods, depending on how they were started). I've done with Kubectl in bash, need to move to Go.

- Test in different scenarios. Error cases / debugging multiple projects etc.. 

- Interface with logic to restart projects in k8s (@rwalle61 working on this simultaneously).

**Comment by rwalle61** [comment]
/assign

I'm working on getting Turbine to restart remote projects in run and debug modes. I've got Turbine to try to restart the projects in these modes (https://github.com/eclipse/codewind/pull/2519) and am now working on the scripts (`nodejs-container.sh` first) to actually restart projects in run and ebug modes.

VSCode (and probably the other IDEs) will not need many changes to allow the user to restart remote projects (https://github.com/eclipse/codewind-vscode/pull/519), because the IDEs automatically adapt to the project capabilities, which I am changing in Turbine

**Comment by rwalle61** [comment]
@jopit @eharris369 @tetchel what debug ports would you like to display in the UI?
![image](https://user-images.githubusercontent.com/18170169/77915421-5c246400-728f-11ea-937a-c814c35f41b2.png)

To take Node projects for example, "Internal Debug Port" would be 9229. `cwctl` will forward that to (e.g.) `localhost:4000`. The user can attach their debugger to `localhost:4000` to debug their remote app.


What should "External Debug Port" be in this example?

**Comment by tetchel** [comment]
We could rename 'exposed debug port' for remote projects to something like 'forwarded debug port' ?

**Comment by jopit** [comment]
I don't think we should hide the 'External Debug Port' since users may want to manually attach a debugger to it.

If the user is likely to get confused about which host the port is on, isn't it a better solution to tell the user which host to use?

**Comment by tetchel** [comment]
it could be a debug url instead of a port, like `localhost:4000`. and we could call it something like 'local debug address'

**Comment by jcockbain** [comment]
To summarize a chat between Portal(@rwalle61) and IDE team(@eharris369, @tetchel, @jopit).

The proposed `cwctl port-forward` command was essentially just wrapping the kube API. It was decided that it was less confusing for a user to interact directly with the kubectl command, in an IDE terminal, as follows:  

1. IDEs will call `PFE projects/{id}/restart directly`, and receive status 202
2. IDEs will find an available port (e.g. 42000)
3. IDEs will use that port to update the `launch.json (or equivalent) config correctly (write port: 42000 instead of address: <> )
4. IDEs will have listened and received the projectRestartResult event from PFE, which includes  `ports: { internalDebugPort: 9229 }` (PFE will emit this event once PFE has restarted the node app with npm run debug on port 9229)
5. IDEs will open a terminal with `k port-forward <podname> 42000:9229`
6. IDEs will launch the debug config, which will successfully attach to the pod's internalDebugPort 9229

**Comment by rwalle61** [comment]
The PFE work will be complete when https://github.com/eclipse/codewind/pull/2519 is merged

**Comment by rwalle61** [comment]
FYI when 2519 is in master you should be able to restart projects in the IDEs by right clicking on them. 

In VSCode this error pops up so will have to be removed
![image](https://user-images.githubusercontent.com/18170169/78678569-0aa85480-78e1-11ea-84ab-9840d23608b3.png)

In VSCode (and probably Eclipse) you will still have to attach the debugger yourself, because the debug config is incorrect for remote projects (change `address` to `port: <localForwardedPort>`)

**Comment by rwalle61** [comment]
#2519 is merged now so PFE master should support remote debug :)

**Comment by sishida** [comment]
Hi, team, For documentation, we currently have this file for debugging: https://www.eclipse.org/codewind/debugging.html

From a user's perspective, does debugging on Kube differ? Thank you!

**Comment by eharris369** [comment]
What happened to the old debug page which talked about how to set up node.js debugging in Eclipse? For node.js on Eclipse it does not attach the debugger automatically, instead it has a dialog to help the user set up a debug session in a chrome based browser. There is no node.js debug support in Eclipse. The old page also recommended to users to disable auto build when debugging Java projects and using hot code replace.

**Comment by eharris369** [comment]
Eclipse PR: https://github.com/eclipse/codewind-eclipse/pull/694

**Comment by sishida** [comment]
@eharris369 Is this the file that you were referring to? https://github.com/eclipse/codewind-docs/blob/0.9.0/docs/_documentations/mdt-eclipse-debugproject.md

**Comment by eharris369** [comment]
@sishida Yes, that is the one. It probably does not need to be as detailed as that but the main information needs to be there:

- The part about disabling auto build for Java projects if they want to use hot code replace while debugging
- The part about debugging initialization code for Java projects
- The information about how Node.js debugging is through a chromium based browser and that the tools just help the user set it up

**Comment by tetchel** [comment]
There are two key differences for remote debug vs local.
1. You must have `kubectl` or `oc` configured to be able to issue commands against the same cluster you have Codewind running on before you can debug a remote project. This is required for 2.
2. Just after the project restarts into debug mode, the plugin will run `kubectl port-forward` to forward a port on the local system to the debug port in the project pod. Then, the plugin connects the debugger to the forwarded local port. 

Note that 1. also applies to the remote container shell feature, also being added in 0.12.

**Comment by tobespc** [comment]
I believe this is all done now

**Comment by tobespc** [comment]
@jopit is this finished from an IDE pov ?

**Comment by micgibso** [comment]
ID complete, removing docs label.
